### PR TITLE
Feature/log out

### PR DESCRIPTION
**Pull request by HeIp**
Implements logout to work, adding mutations

### CODE REVIEW

File: backend/src/app/graphql/mutations/checkUsernameTaken.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import { mutationField, inputObjectType, arg } from 'nexus';
```
Comment (HeIp) [inline backend/src/app/graphql/mutations/checkUsernameTaken.ts:1]: Need confirmation from Veronica that this doesn't work

File: backend/src/app/graphql/mutations/deleteStudent.ts [hunk 2]
```diff
@@ -14,16 +14,19 @@ export const DeleteStudentInput = inputObjectType({
 export const deleteStudent = mutationField('deleteStudent',{
```
Comment (HeIp) [inline backend/src/app/graphql/mutations/deleteStudent.ts:14]: Want to know if it's ok if we rename this file to include plural. As it deletes many.

File: backend/src/app/graphql/objects/index.ts [hunk 3]
```diff
@@ -3,3 +3,7 @@ export * from './Content';
 export * from './blocks';
 export * from './slides';
 export * from './users';
+
+export type BatchPayload = {
+    count: number
+  }
```
Comment (HeIp) [inline backend/src/app/graphql/objects/index.ts:9]: Need an explanation for this, don't think this is used at all ?

File: frontend/src/graphql/mutations/LogOutMutation.ts [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+) {
+  commitMutation(environment, {
+    mutation,
+    variables: {},
+    onCompleted: onSuccess,
+    onError: onFailure,
+  });
+}
+
+export { commit };
```
Comment (HeIp) [inline frontend/src/graphql/mutations/LogOutMutation.ts:25]: Creates Front End mutation for calling backend mutation with call backs for frontend

File: backend/src/app/graphql/mutations/logout.ts [hunk 5]
```diff
@@ -9,10 +9,12 @@ import { Context } from '../context'
...
-  async resolve(_root, context:Context){
-    console.log("here")
-    context.res.clearCookie("jwt")
+export const logOut = mutationField("logOut", {
+  type: 'Boolean',
+  args: {},
+  async resolve(_root, {}, context:Context){
+    // console.log(context.res.cookie());
+    context.res.clearCookie('jwt');
+    return true;
```
Comment (HeIp) [inline backend/src/app/graphql/mutations/logout.ts:18]: For future reference:
This is the code added by David to create log out mutation.
return boolean:true type:value combination when clearing/invalidating the cookie.

File: frontend/src/components/hocs/Sidebar/withSidebar.tsx [hunk 6]
```diff
@@ -9,12 +9,13 @@ import {
...
 import { useHistory, Redirect } from 'react-router-dom';
 import { graphql } from 'babel-plugin-relay/macro';
-import { createFragmentContainer } from 'react-relay';
+import { commitLocalUpdate, commitMutation, createFragmentContainer } from 'react-relay';
 import QueryLookupRenderer from 'relay-query-lookup-renderer';
 import { withSidebar_viewer } from './__generated__/withSidebar_viewer.graphql';
 import { withSidebarQueryResponse } from './__generated__/withSidebarQuery.graphql';
 import { environment } from '../../../graphql/relay';
 import { AppLogo } from '../../icons';
+import {commit as commitLogOutMutation} from '../../../graphql/mutations/LogOutMutation';
```
Comment (HeIp) [inline frontend/src/components/hocs/Sidebar/withSidebar.tsx:18]: here we import logout mutation

File: frontend/src/components/hocs/Sidebar/withSidebar.tsx [hunk 7]
```diff
@@ -9,12 +9,13 @@ import {
 } from '@ant-design/icons';
 import { useHistory, Redirect } from 'react-router-dom';
 import { graphql } from 'babel-plugin-relay/macro';
-import { createFragmentContainer } from 'react-relay';
+import { commitLocalUpdate, commitMutation, createFragmentContainer } from 'react-relay';
```
Comment (HeIp) [inline frontend/src/components/hocs/Sidebar/withSidebar.tsx:12]: hopefully someone can explain what this does, don't remember

File: frontend/src/components/hocs/Sidebar/withSidebar.tsx [hunk 8]
```diff
@@ -43,17 +44,20 @@ const menuItemsForViewer = ({ role }: withSidebar_viewer) => {
       return [
         { name: 'Dashboard', path: '', icon: <DashboardOutlined /> },
         { name: 'Admin', path: 'admin', icon: <LockOutlined /> },
-        { name: 'Lessons', path: 'lessons', icon: <AppstoreOutlined /> },
+        { name: 'Lessons', path: 'lessons', icon: <AppstoreOutlined /> }
```
Comment (HeIp) [inline frontend/src/components/hocs/Sidebar/withSidebar.tsx:47]: Why is this being re-added, when agreed it was to be removed ?

File: frontend/src/components/hocs/Sidebar/withSidebar.tsx [hunk 9]
```diff
@@ -43,17 +44,20 @@ const menuItemsForViewer = ({ role }: withSidebar_viewer) => {
...
       return [
         { name: 'Dashboard', path: '', icon: <DashboardOutlined /> },
         { name: 'Courses', path: 'courses', icon: <BookOutlined /> },
+        // {name: 'Logout', path:'logout', icon: <LogoutOutlined/>}
       ];
     case 'Student':
       return [
         { name: 'Dashboard', path: '', icon: <DashboardOutlined /> },
         { name: 'Courses', path: 'courses', icon: <BookOutlined /> },
+        // {name: 'Logout', path:'logout', icon: <LogoutOutlined/>}
```
Comment (HeIp) [inline frontend/src/components/hocs/Sidebar/withSidebar.tsx:60]: why is {name: 'Logout', path:'logout', icon: <LogoutOutlined/>} commented out ? for all our cases

File: frontend/src/components/hocs/Sidebar/withSidebar.tsx [hunk 10]
```diff
@@ -83,15 +87,26 @@ const Sidebar = createFragmentContainer(
...
     function logOut() {
-      localStorage.setItem('jwt', '');
-      localStorage.removeItem('jwt');
       document.cookie = 'jwt=; expires=Thu, 01 Jan 1970 00:00:00 UTC;';
+      history.push('/login');
+      commitLogOutMutation(
+        logOutSuccess,
+        logOutError
+      );
+
```
Comment (HeIp) [inline frontend/src/components/hocs/Sidebar/withSidebar.tsx:98]: call logout mutation to get rid of them, then push them to login page

**Review by HeIp — COMMENTED** [review]
Comments @verociraptor should review.
And Comments in general for people to look at and maybe answer

File: frontend/src/components/hocs/Sidebar/withSidebar.tsx (hunk 9 above)
Comment (cbarb) [reply frontend/src/components/hocs/Sidebar/withSidebar.tsx:60]: that was one of the temporary fixes i put in to redirect to the login page, can be deleted

**Review by cbarb — APPROVED** [review]

File: frontend/src/components/hocs/Sidebar/withSidebar.tsx (hunk 9 above)
Comment (HeIp) [reply frontend/src/components/hocs/Sidebar/withSidebar.tsx:60]: Ohhh ok, wasn't aware

File: backend/src/app/graphql/mutations/deleteStudent.ts (hunk 2 above)
Comment (verociraptor) [reply backend/src/app/graphql/mutations/deleteStudent.ts:14]: that's fine

File: backend/src/app/graphql/objects/index.ts (hunk 3 above)
Comment (verociraptor) [reply backend/src/app/graphql/objects/index.ts:9]: it's not. was an attempt to fix the resolve for deletestudent

File: frontend/src/components/hocs/Sidebar/withSidebar.tsx (hunk 8 above)
Comment (verociraptor) [reply frontend/src/components/hocs/Sidebar/withSidebar.tsx:47]: might've been an older branch that wasn't updated to that. that can just be removed then.

File: frontend/src/components/hocs/Sidebar/withSidebar.tsx (hunk 7 above)
Comment (verociraptor) [reply frontend/src/components/hocs/Sidebar/withSidebar.tsx:12]: don't think those are being used in the file

File: backend/src/app/graphql/mutations/checkUsernameTaken.ts [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+
+export const TakenUserInput = inputObjectType({
+    name: 'TakenUserInput',
+    definition(t) {
+      t.string('username', {
+        required: true,
+      });
+    },
+  });
+export const checkUsernameTaken = mutationField('checkUsernameTaken',{
```
Comment (verociraptor) [inline backend/src/app/graphql/mutations/checkUsernameTaken.ts:14]: this file should probably be removed since it's not going to be used.

**Review by verociraptor — APPROVED** [review]

**Comment by HeIp** [comment]
Removed that stuff that veronica approved to be killed so here we are along with some linting

File: backend/src/app/server/index.ts [hunk 12]
```diff
@@ -76,7 +76,16 @@ const onServerStart = () =>
...
 
+console.log(NODE_ENV);
 if (NODE_ENV !== 'test') {
-  app.listen({ port: PORT}, onServerStart);
+  console.log('listening');
+  app.listen({ port: PORT || 3000 }, onServerStart);
 }
+
+
+// app.get('/logout', (req, res) => {
```
Comment (andres-valdes) [inline backend/src/app/server/index.ts:86]: every time you push a commented line of code, a puppy dies

Comment (HeIp) [reply backend/src/app/server/index.ts:86]: That's them sir, I make appropriate branches for the features

**Review by jessica — APPROVED** [review]

**Pull request merged**